### PR TITLE
[ENH] Add CovariateRegressor class

### PR DESCRIPTION
**Pull request by chiuhoward**
With some assistance from @richford

- Implements CovariateRegressor for regressing covariates from features
- Includes find_subset_indices utility for hash and precise matching
- Adds IdentityTransformer helper class
- Tests cover edge cases and error handling
- Supports cross-validation, missing value imputation, and flexible pipelines
- All tests pass with full coverage of core functionality

### CODE REVIEW

**Comment by arokem** [comment]
I believe the test failure is unrelated. It might be an update to pytorch API that I need to investigate, before we can have your code run cleanly.

**Comment by chiuhoward** [comment]
@arokem I believe that it’s because `verbose` was deprecated from `ReduceLROnPlateau` at some point? I encountered the same failing tests earlier in the year when I tried to just fix a typo in the docs.

https://github.com/pytorch/pytorch/pull/111302

**Comment by arokem** [comment]
@chiuhoward : do you want to make a separate PR just to fix that failure first, and then we can move on with this one? 

BTW: I gave it a first read through and the code in this PR looks great. I will give it a thorough review once we resolve the test failures, if that's OK with you.

**Comment by arokem** [comment]
OK - now that #29 is merged, you can either rebase this one on top of main, or merge main into this branch and then push again, so we can see that all tests are passing with that fix.

File: afqinsight/covariate_regressor.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            try:
+                indices.append(full_hashes.index(subset_hash))
+            except ValueError as e:
+                raise ValueError(f"No matching row found for subset row {i}") from e
+    else:
+        raise ValueError(f"Unknown method '{method}'. Use 'hash' or 'precise'.")
+    return np.array(indices)
+
+
+class IdentityTransformer(BaseEstimator, TransformerMixin):
```
Comment (arokem) [inline afqinsight/covariate_regressor.py]: What is this class for?

File: afqinsight/covariate_regressor.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    indices = []
+    if method == "precise":
+        for i, subset_row in enumerate(X_subset):
+            matches = [
+                j
+                for j, full_row in enumerate(X_full)
+                if np.array_equal(full_row, subset_row, equal_nan=True)
+            ]
+            if not matches:
+                raise ValueError(f"No matching row found for subset row {i}")
```
Comment (arokem) [inline afqinsight/covariate_regressor.py:26]: Is erroring out really what we want to happen when there are no matches? Maybe return an empty array instead?

File: afqinsight/covariate_regressor.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
+import numpy as np
+from scipy.linalg import lstsq
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.impute import SimpleImputer
+
+
+def find_subset_indices(X_full, X_subset, method="hash"):
```
Comment (arokem) [inline afqinsight/covariate_regressor.py]: I wonder if there is a pure numpy way of doing this without looping over elements (e.g., through some tiling). Couldn't come up with it with a bit of thought, though, so maybe there isn't something like that.

File: afqinsight/CovariateRegressor.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            Array of length (n_samples, n_features), from which the covariate
+            will be regressed. This is used to determine how the
+            covariate-models should be cross-validated (which is necessary
+            to use in in scikit-learn Pipelines).
+        cross_validate : bool
+            Whether to cross-validate the covariate-parameters (y~covariate)
+            estimated from the train-set to the test set (cross_validate=True)
+            or whether to fit the covariate regressor separately on the test-set
+            (cross_validate=False). Setting this parameter to True is equivalent
+            to "foldwise covariate regression" (FwCR) as described in our paper
```
Comment (arokem) [inline afqinsight/CovariateRegressor.py]:
```suggestion
            to "foldwise covariate regression" (FwCR) as described in Snoek et al. 
```

And add a reference in the Notes section. 

BTW: if you used code from Snoek et al.'s previous work that's probably fine, but we might need to add their original license somewhere, to comply with that license's requirements. Do you have a pointer to the original code?

File: afqinsight/CovariateRegressor.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            (https://www.biorxiv.org/content/early/2018/03/28/290684). Setting
+            this parameter to False, however, is NOT equivalent to "whole
+            dataset covariate regression" (WDCR) as it does not apply covariate
+            regression to the *full* dataset, but simply refits the covariate
+            model on the test-set. We recommend setting this parameter to True.
+        precise: bool
+            Transformer-objects in scikit-learn only allow to pass the data
```
Comment (arokem) [inline afqinsight/CovariateRegressor.py]: The explanations of the parameters are rather long. I wonder whether a "Notes" section would better accommodate these explanations?

File: afqinsight/covariate_regressor.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        ----------
+        X : numpy array
+            An array of shape (n_samples, n_features), which should correspond
+            to your train-set only!
+        y : None
+            Included for compatibility; does nothing.
+        """
+
+        # Prepare covariate matrix (adds intercept if needed)
+        covariate = self._prepare_covariate(self.covariate)
```
Comment (arokem) [inline afqinsight/covariate_regressor.py:150]: I think you can replace this with and get rid of the `_prepare_covariate` method, which is essentially just this if statement:
```suggestion
        if self.stack_intercept:
            covariate = np.c_[np.ones((covariate.shape[0], 1)), covariate]
```

Or am I missing something?

File: afqinsight/covariate_regressor.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        self.X_imputer = SimpleImputer(strategy="median")
+        self.unique_id_col_index = unique_id_col_index
+
+    def _prepare_covariate(self, covariate):
+        """Prepare covariate matrix (adds intercept if needed)"""
+        if self.stack_intercept:
+            return np.c_[np.ones((covariate.shape[0], 1)), covariate]
+        return covariate
+
+    def fit(self, X, y=None):
```
Comment (arokem) [inline afqinsight/covariate_regressor.py:137]: I find it confusing that there is an input X here, but also an `self.X`, which got assigned on initialization. Would it be possible to name them something that differentiates between them (e.g., `X_covariates` and `X_train`) or something like that?

File: afqinsight/covariate_regressor.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+class CovariateRegressor(BaseEstimator, TransformerMixin):
+    """
+    Fits covariate(s) onto each feature in X and returns their residuals.
+    """
+
+    def __init__(
+        self,
+        covariate,
+        X,
+        pipeline=None,
```
Comment (arokem) [inline afqinsight/covariate_regressor.py:53]: The `pipeline` parameter is not documented in the docstring.

**Review by arokem — COMMENTED** [review]
I have not read through the tests yet, but here are a few comments for now.

File: afqinsight/CovariateRegressor.py (hunk 4 above)
Comment (chiuhoward) [reply afqinsight/CovariateRegressor.py]: https://github.com/lukassnoek/MVCA?tab=MIT-1-ov-file#readme

This is the original repository and the original license is a MIT one

Comment (arokem) [reply afqinsight/CovariateRegressor.py]: Gotcha! As you can see, the license states that: 

```
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
```

I would add that either in our license file. Something like: 

```
Code used in the covariate_regressor module is distributed with the following license: 
```

And then you past their entire license after it.
 
BTW: I would also change the file-name to `covariate_regressor.py`, because it is not usual to have Python module names with CamelCase.

Comment (arokem) [reply afqinsight/CovariateRegressor.py]: Or maybe `covariate_regression.py`?

File: afqinsight/covariate_regressor.py (hunk 1 above)
Comment (chiuhoward) [reply afqinsight/covariate_regressor.py]: In `afqinsight.pipeline.make_base_afq_pipeline`, I wanted to pass in a feature transformer that already had a scaler and imputer in it. So I had to turn off the `scaler` (default: standard) and `imputer` (default: simple) in the function. However, if you pass `None` to `make_base_afq_pipeline` for the `scaler` and `imputer` parameters, then it falls back to the defaults rather than omitting that step. So the `IdentityTransformer` was a workaround to bypass that default. But you’re right that it doesn’t belong in AFQ-Insight so I’ll remove it from this PR.

File: afqinsight/covariate_regressor.py (hunk 2 above)
Comment (chiuhoward) [reply afqinsight/covariate_regressor.py:26]: We think an error should be thrown at some point, but have provided an additional kwarg for people to insist on proceeding with `allow_missing`.

File: afqinsight/covariate_regressor.py (hunk 3 above)
Comment (chiuhoward) [reply afqinsight/covariate_regressor.py]: It wasn’t immediately apparent to us but our skills may be lacking :’)

File: afqinsight/CovariateRegressor.py (hunk 4 above)
Comment (chiuhoward) [reply afqinsight/CovariateRegressor.py]: Renamed files and appended suggestion to AFQ-Insight’s license file below our BSD-3 license

File: afqinsight/covariate_regressor.py (hunk 6 above)
Comment (chiuhoward) [reply afqinsight/covariate_regressor.py:150]: We repeat this in both fit and transform, so we thought it was a good idea to keep this as a separate DRY function.

File: afqinsight/covariate_regressor.py (hunk 7 above)
Comment (chiuhoward) [reply afqinsight/covariate_regressor.py:137]: I think there is a requirement for `X` and `y` in the `fit` and `transform` functions to be called exactly that, so I have renamed `self.X` to `self.X_full`

File: afqinsight/covariate_regressor.py (hunk 1 above)
Comment (chiuhoward) [reply afqinsight/covariate_regressor.py]: I actually had to move it to the tests, but please let me know where might be a better place to put this/how we can work around this completely if possible!

Comment (arokem) [reply afqinsight/covariate_regressor.py]: OK - thanks! That's helpful to understand. 

I see this as a limitation of the functionality of `make_base_afq_pipeline`, because your use-case is completely reasonable and is not addressed with current functionality. That said, I wonder if what we want is to implement this functionality as part of what `make_base_afq_pipeline` does. 

That said, I'd advocate for removing IdentityTransformer from this module and also from the tests. If anything I think this should go somewhere in afqinsight/pipeline.py and tested in test_pipeline.py. 

Does that make sense to you? We can also do that on a separate PR!

Comment (chiuhoward) [reply afqinsight/covariate_regressor.py]: To keep this PR atomic, I will rework this PR to remove any usage of `IdentityTransformer`, then build that in a separate PR!

**Comment by arokem** [comment]
Thanks for your work on this @chiuhoward!
